libs/core/algorithms/include/hpx/parallel/unseq/loop.hpp

```diff
@@ -0,0 +1,356 @@
+            }
+        };
+
+        ///////////////////////////////////////////////////////////////////////
+        struct unseq_loop
+        {
+            template <typename Begin, typename End, typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE static typename std::enable_if<
+                hpx::traits::is_random_access_iterator_v<Begin>, Begin>::type
+            call(Begin it, End end, F&& f)
+            {
+                return unseq_loop_n::call(
+                    it, std::distance(it, end), HPX_FORWARD(F, f));
+            }
+
+            template <typename Begin, typename End, typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE static typename std::enable_if<
+                !hpx::traits::is_random_access_iterator_v<Begin>, Begin>::type
+            call(Begin it, End end, F&& f)
+            {
+                for (/* */; it != end; it++)
+                {
+                    HPX_INVOKE(f, it);
+                }
+            }
+
+            template <typename Begin, typename End, typename CancelToken,
+                typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE static typename std::enable_if<
+                hpx::traits::is_random_access_iterator_v<Begin>, Begin>::type
+            call(Begin it, End end, CancelToken& tok,
+                F&& f)
+            {
+                return unseq_loop_n::call(
+                    it, std::distance(it, end), tok, HPX_FORWARD(F, f));
+            }
+
+            template <typename Begin, typename End, typename CancelToken,
+                typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE static typename std::enable_if<
+                !hpx::traits::is_random_access_iterator_v<Begin>, Begin>::type
+            call(Begin it, End end, CancelToken& tok,
+                F&& f)
+            {
+                for (/* */; it != end; ++it)
+                {
+                    HPX_INVOKE(f, it);
+                    if (tok.was_cancelled())
+                        return it;
+                }
+                return it;
+            }
+        };
+
+        struct unseq_loop_ind
+        {
+            template <typename Begin, typename End, typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE static typename std::enable_if<
+                hpx::traits::is_random_access_iterator_v<Begin>, Begin>::type
+            call(Begin it, End end, F&& f)
+            {
+                return unseq_loop_n_ind::call(
+                    it, std::distance(it, end), HPX_FORWARD(F, f));
+            }
+
+            template <typename Begin, typename End, typename F>
+            HPX_HOST_DEVICE HPX_FORCEINLINE static typename std::enable_if<
+                !hpx::traits::is_random_access_iterator_v<Begin>, Begin>::type
+            call(Begin it, End end, F&& f)
+            {
+                for (/* */; it != end; ++it)
+                {
+                    HPX_INVOKE(f, *it);
+                }
+                return it;
+            }
```

Comment:
Could we make this fall back to the existing non-vectorized implementation? This comment applies to many more places in this file below.